### ISSUE TITLE
Why is the training in MS-CMRseg2019 different from that Bladder-Tumor-seg?

### ISSUE THREAD

**Issue by pangda72** [LynnHg/Pytorch-medical-image-segmentation#3]
Hi,@LynnHg   

I am sorry to bother you .I am a beginner in image segmentation.I found that train.py in [MS-CMRseg2019](https://github.com/LynnHg/Pytorch-medical-image-segmentation/blob/main/train/train_mscmr2019.py) different from that [Bladder-Tumor-seg.](https://blog.csdn.net/baidu_36511315/article/details/106106463) I felt confused.Could you give me some hints?

eg:
 In the MS-CMRseg2019:

            X1 = input.cuda()
            y = mask.cuda()
            optimizer.zero_grad()
            output = net(X1)
            output = torch.sigmoid(output)
            loss = criterion(output, y)
            loss.backward()
            optimizer.step()
            train_losses.append(loss.item())
            class_dice = []
            for i in range(1, mscmr2019.num_classes):
                cur_dice = diceCoeffv2(output[:, i:i + 1, :], y[:, i:i + 1, :]).cpu().item()
                class_dice.append(cur_dice)

But in the [Bladder-Tumor-seg:](https://blog.csdn.net/baidu_36511315/article/details/106106463)

            X = inputs.cuda()
            y = mask.cuda()
            optimizer.zero_grad()
            output = net(X)
            loss = criterion(output, y)
            output = torch.sigmoid(output)
            output[output < 0.5] = 0
            output[output > 0.5] = 1
            loss.backward()
            optimizer.step()
            bladder_dice = diceCoeffv2(output[:, 0:1, :], y[:, 0:1, :], activation=None).cpu().item()
            tumor_dice = diceCoeffv2(output[:, 1:2, :], y[:, 1:2, :], activation=None).cpu().item()
            mean_dice = (bladder_dice + tumor_dice) / 2

Where I'm confused:

(1)In the Bladder-Tumor-seg, output[output < 0.5] = 0 output[output > 0.5] = 1 is required before calculating dice values,But it's not found  in the MS-CMRseg2019.I don't know if I need these two lines of code.Could give me some help?

(2) In the MS-CMRseg2019: for i in range(1,num_classes),but In the Bladder-Tumor-seg: for i in range(0,num_classes).You mean the first category is background In the MS-CMRseg2019,but  In the Bladder-Tumor-seg, the last category is background.But these two repo,it's the same in the palette, the same in the one-hot code,I don't know which part of the program would represent the category of background,Could you give me some hints?

Thanks in advance.
Looking forward to hearing from you!
Best.

**Comment by LynnHg** [comment]
> Hi,@LynnHg
> 
> I am sorry to bother you .I am a beginner in image segmentation.I found that train.py in [MS-CMRseg2019](https://github.com/LynnHg/Pytorch-medical-image-segmentation/blob/main/train/train_mscmr2019.py) different from that [Bladder-Tumor-seg.](https://blog.csdn.net/baidu_36511315/article/details/106106463) I felt confused.Could you give me some hints?
> 
> eg: In the MS-CMRseg2019:
> 
> ```
>         X1 = input.cuda()
>         y = mask.cuda()
>         optimizer.zero_grad()
>         output = net(X1)
>         output = torch.sigmoid(output)
>         loss = criterion(output, y)
>         loss.backward()
>         optimizer.step()
>         train_losses.append(loss.item())
>         class_dice = []
>         for i in range(1, mscmr2019.num_classes):
>             cur_dice = diceCoeffv2(output[:, i:i + 1, :], y[:, i:i + 1, :]).cpu().item()
>             class_dice.append(cur_dice)
> ```
> 
> But in the [Bladder-Tumor-seg:](https://blog.csdn.net/baidu_36511315/article/details/106106463)
> 
> ```
>         X = inputs.cuda()
>         y = mask.cuda()
>         optimizer.zero_grad()
>         output = net(X)
>         loss = criterion(output, y)
>         output = torch.sigmoid(output)
>         output[output < 0.5] = 0
>         output[output > 0.5] = 1
>         loss.backward()
>         optimizer.step()
>         bladder_dice = diceCoeffv2(output[:, 0:1, :], y[:, 0:1, :], activation=None).cpu().item()
>         tumor_dice = diceCoeffv2(output[:, 1:2, :], y[:, 1:2, :], activation=None).cpu().item()
>         mean_dice = (bladder_dice + tumor_dice) / 2
> ```
> 
> Where I'm confused:
> 
> (1)In the Bladder-Tumor-seg, output[output < 0.5] = 0 output[output > 0.5] = 1 is required before calculating dice values,But it's not found in the MS-CMRseg2019.I don't know if I need these two lines of code.Could give me some help?
> 
> (2) In the MS-CMRseg2019: for i in range(1,num_classes),but In the Bladder-Tumor-seg: for i in range(0,num_classes).You mean the first category is background In the MS-CMRseg2019,but In the Bladder-Tumor-seg, the last category is background.But these two repo,it's the same in the palette, the same in the one-hot code,I don't know which part of the program would represent the category of background,Could you give me some hints?
> 
> Thanks in advance. Looking forward to hearing from you! Best.

1)You can refer to the latest [Bladder-Tumor-seg](https://blog.csdn.net/baidu_36511315/article/details/120902937?spm=1001.2014.3001.5501), and the new code in this article can be downloaded and run.
2)output[output < 0.5] = 0 output[output > 0.5] = 1 is a kind of post-processing, not necseeary.

**Comment by pangda72** [comment]
> > Hi,@LynnHg
> > I am sorry to bother you .I am a beginner in image segmentation.I found that train.py in [MS-CMRseg2019](https://github.com/LynnHg/Pytorch-medical-image-segmentation/blob/main/train/train_mscmr2019.py) different from that [Bladder-Tumor-seg.](https://blog.csdn.net/baidu_36511315/article/details/106106463) I felt confused.Could you give me some hints?
> > eg: In the MS-CMRseg2019:
> > ```
> >         X1 = input.cuda()
> >         y = mask.cuda()
> >         optimizer.zero_grad()
> >         output = net(X1)
> >         output = torch.sigmoid(output)
> >         loss = criterion(output, y)
> >         loss.backward()
> >         optimizer.step()
> >         train_losses.append(loss.item())
> >         class_dice = []
> >         for i in range(1, mscmr2019.num_classes):
> >             cur_dice = diceCoeffv2(output[:, i:i + 1, :], y[:, i:i + 1, :]).cpu().item()
> >             class_dice.append(cur_dice)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But in the [Bladder-Tumor-seg:](https://blog.csdn.net/baidu_36511315/article/details/106106463)
> > ```
> >         X = inputs.cuda()
> >         y = mask.cuda()
> >         optimizer.zero_grad()
> >         output = net(X)
> >         loss = criterion(output, y)
> >         output = torch.sigmoid(output)
> >         output[output < 0.5] = 0
> >         output[output > 0.5] = 1
> >         loss.backward()
> >         optimizer.step()
> >         bladder_dice = diceCoeffv2(output[:, 0:1, :], y[:, 0:1, :], activation=None).cpu().item()
> >         tumor_dice = diceCoeffv2(output[:, 1:2, :], y[:, 1:2, :], activation=None).cpu().item()
> >         mean_dice = (bladder_dice + tumor_dice) / 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Where I'm confused:
> > (1)In the Bladder-Tumor-seg, output[output < 0.5] = 0 output[output > 0.5] = 1 is required before calculating dice values,But it's not found in the MS-CMRseg2019.I don't know if I need these two lines of code.Could give me some help?
> > (2) In the MS-CMRseg2019: for i in range(1,num_classes),but In the Bladder-Tumor-seg: for i in range(0,num_classes).You mean the first category is background In the MS-CMRseg2019,but In the Bladder-Tumor-seg, the last category is background.But these two repo,it's the same in the palette, the same in the one-hot code,I don't know which part of the program would represent the category of background,Could you give me some hints?
> > Thanks in advance. Looking forward to hearing from you! Best.
> 
> 1)You can refer to the latest [Bladder-Tumor-seg](https://blog.csdn.net/baidu_36511315/article/details/120902937?spm=1001.2014.3001.5501), and the new code in this article can be downloaded and run. 2)output[output < 0.5] = 0 output[output > 0.5] = 1 is a kind of post-processing, not necseeary.

Hi, @LynnHg 
Thank you for your help.I feel very happy.I will try as you suggest.
Best.
